myia/pipeline.py

```diff
@@ -0,0 +1,516 @@
+        )
+
+    def make(self):
+        """Create a Pipeline from this definition."""
+        return Pipeline(self)
+
+    def __getitem__(self, item):
+        """Return a pipeline that only contains a subset of the steps."""
+        steps = list(self.steps.items())
+        return PipelineDefinition(
+            resources=self.resources,
+            steps=dict(steps[self.getslice(item)])
+        )
+
+
+class Pipeline:
+    """Sequence of connected processing steps.
+
+    Created from a PipelineDefinition. Each step processes the output of the
+    previous step. A Pipeline also defines common resources for all the steps.
+    """
+
+    def __init__(self, defn):
+        """Initialize a Pipeline from a PipelineDefinition."""
+        def convert(x, name):
+            if isinstance(x, Partial):
+                x = x.partial(pipeline_init={'pipeline': self, 'name': name})
+                x = x()
+            return x
+
+        self.defn = defn
+        self.resources = NS()
+        self.steps = NS()
+        self._seq = []
+        for k, v in defn.resources.items():
+            self.resources[k] = convert(v, None)
+        for k, v in defn.steps.items():
+            v = convert(v, k)
+            self.steps[k] = v
+            self._seq.append(v)
+
+    def __getitem__(self, item):
+        """Return a sub-pipeline.
+
+        item may be the name of a step, in which case the resulting
+        sub-pipeline goes from the beginning to that step (inclusive), or it
+        may be a range that starts at a given step and ends at another
+        (again, inclusive).
+        """
+        return _PipelineSlice(self, item)
+
+    def __call__(self, **args):
+        """Execute the pipeline from start to finish."""
+        return self[:](**args)
+
+
+class _PipelineSlice:
+    def __init__(self, pipeline, slice):
+        self.pipeline = pipeline
+        self.slice = self.pipeline.defn.getslice(slice)
+
+    def __call__(self, **args):
+        for step in self.pipeline._seq[self.slice]:
+            if step.active:
+                valid_args, rest = _filter_keywords(step.step, args)
+                args = {**args, **step.step(**valid_args)}
+        return args
+
+
+class PipelineStep(Partializable):
+    """Step in a Pipeline."""
+
+    def __init__(self, pipeline_init):
+        """Initialize a PipelineStep."""
+        self.active = pipeline_init.get('active', True)
+        self.name = pipeline_init['name']
+        self.pipeline = pipeline_init['pipeline']
+        self.steps = self.pipeline.steps
+        self.resources = self.pipeline.resources
+
+    def step(self, **kwargs):
+        """Execute this step only.
+
+        The received arguments come from the previous step.
+        """
+        raise NotImplementedError()
+
+    def __call__(self, **args):
+        """Execute the pipeline all the way up to this step."""
+        return self.pipeline[:self.name](**args)
```

Comment:
I'm not sure how would someone call this.

Reply:
If the steps are `parse,resolve,export`, `pipeline.steps.resolve(input=xyz)` would run `parse,resolve` and stop there. That interface was used at some point to call the `convert` step, when it was a step, but maybe it's not relevant anymore.